src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs

```diff
@@ -659,6 +659,67 @@ public MethodIL GetMethodILWithInlinedSubstitutions(MethodIL method)
             return new SubstitutedMethodIL(method.GetMethodILDefinition(), newBody, newEHRegions.ToArray(), debugInfo, newStrings.ToArray());
         }
 
+        private bool TryGetMethodConstantValue(MethodDesc method, out int constant, int level = 0)
+        {
+            method = method.GetTypicalMethodDefinition();
+
+            TypeFlags returnType = method.Signature.ReturnType.Category;
+            if (returnType is < TypeFlags.Boolean or > TypeFlags.UInt32
+                || method.IsIntrinsic
+                || _nestedILProvider.GetMethodIL(method) is not MethodIL methodIL)
+            {
+                constant = 0;
+                return false;
+            }
+
+            var reader = new ILReader(methodIL.GetILBytes());
+            var opcode = reader.ReadILOpcode();
+            switch (opcode)
+            {
+                case ILOpcode.ldc_i4: constant = (int)reader.ReadILUInt32(); break;
+                case ILOpcode.ldc_i4_s: constant = (sbyte)reader.ReadILByte(); break;
+                case >= ILOpcode.ldc_i4_0 and <= ILOpcode.ldc_i4_8: constant = opcode - ILOpcode.ldc_i4_0; break;
+                case ILOpcode.ldc_i4_m1: constant = -1; break;
```

Comment:
This is handling direct constants. Are indirect constants also desirable to be handled or are they possible handled elsewhere?

Consider something like `public static float X => DegreesToRadians(90);`, where the `DegreesToRadians(90)` call can be trivially folded and so `X` is effectively producing a constant

Reply:
I considered just hooking this into the cctor interpreter, but I don't know if I want to go there. This runs on all IL and if it gets too fancy, we might see it on throughput.

We should ideally not have to do this rewriting - we should use RyuJIT as the IL scanner because RyuJIT already can do this (and more) better.